api/api/admin/forms.py

```diff
@@ -0,0 +1,33 @@
+from django import forms
+
+from api.models import UserPreferences
+
+
+class UserPreferencesAdminForm(forms.ModelForm):
+    blur_images = forms.BooleanField(initial=True, required=False)
+
+    class Meta:
+        model = UserPreferences
+        fields = ["blur_images"]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.initial["blur_images"] = self._get_blur_images()
+
+    def _get_blur_images(self):
+        mod_preferences = self.instance.preferences.get("moderator", {})
+        return mod_preferences.get("blur_images", True)
+
+    @staticmethod
+    def _set_blur_images(instance, value):
+        mod_preferences = instance.preferences.get("moderator", {})
+        mod_preferences["blur_images"] = value
+        instance.preferences = instance.preferences | {"moderator": mod_preferences}
```

Comment:
Theoretically, this is mixing concerns a bit between the form and the model. The logic of how to read and save data on the model should live in the model, with the form ideally able to do simple gets and sets on model attributes. That's also true for the correct default values of any part of the model's data, that should all live on the model class.

Because this is a "virtual" field, we need to implement it on the model with `@property`, but it lets us move the logic for getting and setting, including expected defaults, all onto the model, and reduce the form's concerns. There's also some unnecessary code in the `save` method of the form we can remove:

```diff
diff --git a/api/api/admin/forms.py b/api/api/admin/forms.py
index 48a43426c..2ac566773 100644
--- a/api/api/admin/forms.py
+++ b/api/api/admin/forms.py
@@ -12,22 +12,8 @@ class UserPreferencesAdminForm(forms.ModelForm):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.initial["blur_images"] = self._get_blur_images()
-
-    def _get_blur_images(self):
-        mod_preferences = self.instance.preferences.get("moderator", {})
-        return mod_preferences.get("blur_images", True)
-
-    @staticmethod
-    def _set_blur_images(instance, value):
-        mod_preferences = instance.preferences.get("moderator", {})
-        mod_preferences["blur_images"] = value
-        instance.preferences = instance.preferences | {"moderator": mod_preferences}
+        self.initial["blur_images"] = self.instance.blur_images
 
     def save(self, commit=True):
-        instance: UserPreferences = super().save(commit=False)
-        blur_images = self.cleaned_data.get("blur_images", True)
-        self._set_blur_images(instance, blur_images)
-        if commit:
-            instance.save()
-        return instance
+        self.instance.blur_images = self.cleaned_data.get("blur_images")
+        return super().save(commit=commit)
diff --git a/api/api/models/moderation.py b/api/api/models/moderation.py
index bf4677fb8..fc818a239 100644
--- a/api/api/models/moderation.py
+++ b/api/api/models/moderation.py
@@ -11,6 +11,22 @@ class UserPreferences(models.Model):
     def __str__(self):
         return f"{self.user.username}'s preferences"
 
+    @property
+    def moderator(self):
+        return self.preferences.get("moderator", {})
+
+    @moderator.setter
+    def moderator(self, value):
+        self.preferences["moderator"] = value
+
+    @property
+    def blur_images(self):
+        return self.moderator.get("blur_images", True)
+
+    @blur_images.setter
+    def blur_images(self, value):
+        self.moderator |= {"blur_images": value}
+
 
 @receiver(post_save, sender=User)
 def create_or_update_user_profile(sender, instance, created, **kwargs):
```

It's all the same relative complexity, but moves it to the "right" place, with the goal of making it possible for any code to read `blur_images` from the model with the exact same behaviour _vis-a-vis_ the handling of retrival off the model's data and correct defaults. Otherwise, places that referenced the model would need to re-implement the same retrieval and default handling logic the form currently implements.

Reply:
This makes a _ton_ of sense, thanks for taking the time to explain it to me!! It's interesting that there are so many ways to accomplish this, but my own infrequency in working with Django means that I'm not as clued into thinking this way by default. These changes seem a lot better, I've incorporated them 🚀